### PR TITLE
[CodeGen] Use SmallVector for MBB preds/succs

### PR DESCRIPTION
**Pull request by aengelke**
Avoid extra heap allocations for typical predecessor/successor counts.

[c-t-t](http://llvm-compile-time-tracker.com/compare.php?from=e7f9d8e5c3e49e729c69aaa9be3322f7902370b8&to=61a0ed0ea4eab2a9d4b0f1d1ce7b62b57b7cf50e&stat=instructions:u) -0.14% stage2-O0-g

### CODE REVIEW

**Comment by nikic** [comment]
```FAILED: lib/CodeGen/CMakeFiles/LLVMCodeGen.dir/MIRSampleProfile.cpp.obj
sccache C:\BuildTools\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\cl.exe  /nologo /TP -DBUILD_EXAMPLES -DGTEST_HAS_RTTI=0 -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_GLIBCXX_ASSERTIONS -D_HAS_EXCEPTIONS=0 -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Ilib\CodeGen -IC:\ws\src\llvm\lib\CodeGen -Iinclude -IC:\ws\src\llvm\include /DWIN32 /D_WINDOWS   /Zc:inline /Zc:preprocessor /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd5105 -wd4324 -w14062 -we4238 /Gw /O2 /Ob2  -MD  /EHs-c- /GR- -UNDEBUG -std:c++17 /showIncludes /Folib\CodeGen\CMakeFiles\LLVMCodeGen.dir\MIRSampleProfile.cpp.obj /Fdlib\CodeGen\CMakeFiles\LLVMCodeGen.dir\LLVMCodeGen.pdb /FS -c C:\ws\src\llvm\lib\CodeGen\MIRSampleProfile.cpp
C:\ws\src\llvm\lib\CodeGen\MIRSampleProfile.cpp(136): error C2440: 'return': cannot convert from 'llvm::iterator_range<llvm::MachineBasicBlock::pred_iterator>' to 'llvm::iterator_range<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>'
        with
        [
            _Ty=llvm::MachineBasicBlock *
        ]
C:\ws\src\llvm\lib\CodeGen\MIRSampleProfile.cpp(136): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
C:\ws\src\llvm\lib\CodeGen\MIRSampleProfile.cpp(139): error C2440: 'return': cannot convert from 'llvm::iterator_range<llvm::MachineBasicBlock::pred_iterator>' to 'llvm::iterator_range<std::_Vector_iterator<std::_Vector_val<std::_Simple_types<_Ty>>>>'
        with
        [
            _Ty=llvm::MachineBasicBlock *
        ]
C:\ws\src\llvm\lib\CodeGen\MIRSampleProfile.cpp(139): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
```

File: llvm/include/llvm/Analysis/RegionInfoImpl.h [hunk 1]
```diff
@@ -814,7 +814,7 @@ RegionInfoBase<Tr>::getMaxRegionExit(BlockT *BB) const {
     // Get the single exit of BB.
     if (R && R->getEntry() == BB)
       Exit = R->getExit();
-    else if (++BlockTraits::child_begin(BB) == BlockTraits::child_end(BB))
+    else if (BlockTraits::child_begin(BB) + 1 == BlockTraits::child_end(BB))
```
Comment (nikic) [inline llvm/include/llvm/Analysis/RegionInfoImpl.h:817]: Use std::next() to make this independent of implementation details?

**Review by arsenm — COMMENTED** [review]
I'm surprised this wasn't already the case

**Review by nikic — APPROVED** [review]
LGTM

**Review by arsenm — APPROVED** [review]

**Review by MaskRay — APPROVED** [review]

**Pull request merged**